sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -1247,33 +1315,107 @@ public void startBundle(StartBundleContext c) throws IOException {
             factory.getServiceForWriting(id, config, writeOptions, c.getPipelineOptions());
         bigtableWriter = serviceEntry.getService().openForWriting(writeOptions);
       }
+
+      badRecords = new HashSet<>();
+      unbatchedWrites = new HashSet<>();
     }
 
     @ProcessElement
     public void processElement(ProcessContext c, BoundedWindow window) throws Exception {
       checkForFailures();
       KV<ByteString, Iterable<Mutation>> record = c.element();
-      bigtableWriter
-          .writeRecord(record)
-          .whenComplete(
-              (mutationResult, exception) -> {
-                if (exception != null) {
-                  failures.add(new BigtableWriteException(record, exception));
-                }
-              });
+      bigtableWriter.writeRecord(record).whenComplete(handleMutationException(record, window));
       ++recordsWritten;
       seenWindows.compute(window, (key, count) -> (count != null ? count : 0) + 1);
     }
 
+    public BiConsumer<MutateRowResponse, Throwable> handleMutationException(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      return (MutateRowResponse result, Throwable exception) -> {
+        if (exception != null) {
+          if (exception instanceof NotFoundException
+              && !((NotFoundException) exception).isRetryable()) {
+            // This case, of being an NotFoundException and not retryable,
```

Comment:
Would you mind explain that in what kind of scenario (in practice) this (`exception instanceof NotFoundException && !((NotFoundException) exception).isRetryable()`)could happen, its likelihood, and thus evaluate the possible performance implementations.

Reply:
yep, expanded

Reply:
NotFound should never be retriable. Also can you extract out the predicate logic to determine is the result/exception is a bad record?

Reply:
I dont think NotFound is the only exception type that constitutes a bad record. Other possibilities:
- entry w/o mutations
- empty row keys
- timestamps < -1
- qualifiers > 16KiB

Those should generate Invalid Arg exceptions

Reply:
I've added these test cases.